### PR TITLE
Fix #3205 MOM4-JinHeeYuk-ocean_salt

### LINKED ISSUE #3205: MOM4-JinHeeYuk-ocean_salt IsoSurface and volume blank
MOM4-JinHeeYuk-ocean_salt-IsoSurface
[MOM4-JinHeeYuk-ocean_salt-IsoSurface.png.log](https://github.com/NCAR/VAPOR/files/9441951/MOM4-JinHeeYuk-ocean_salt-IsoSurface.png.log)
[MOM4-JinHeeYuk-ocean_salt-IsoSurface.png.vs3.zip](https://github.com/NCAR/VAPOR/files/9441952/MOM4-JinHeeYuk-ocean_salt-IsoSurface.png.vs3.zip)
MOM4-JinHeeYuk-ocean_salt-Volume
[MOM4-JinHeeYuk-ocean_salt-Volume.png.log](https://github.com/NCAR/VAPOR/files/9441953/MOM4-JinHeeYuk-ocean_salt-Volume.png.log)
[MOM4-JinHeeYuk-ocean_salt-Volume.png.vs3.zip](https://github.com/NCAR/VAPOR/files/9441954/MOM4-JinHeeYuk-ocean_salt-Volume.png.vs3.zip)

### PR DESCRIPTION
**Pull request by StasJ**
This dataset was triggering issues with different OpenGL driver limitations depending on the system. This PR adds proper error handling for these scenarios. It also adds a mechanism for predicting the amount of GPU RAM required to render a dataset and lets the user know right away if they don't have enough, rather than making them wait for the GPU RAM to overflow and get an error (only works for vendors with extensions for querying GPU RAM since this is not supported in the OpenGL spec). Finally, it improves OpenGL error reporting overall by replacing the old `OpenGL Error #1234` messages with proper error strings.

Fix #3205

### CODE REVIEW

File: lib/render/Texture.cpp [hunk 1]
```diff
@@ -52,6 +53,23 @@ int Texture::TexImage(int internalFormat, int width, int height, int depth, unsi
     VAssert(Initialized());
     VAssert(_nDims >= 2 || height == 0);
     VAssert(_nDims >= 3 || depth == 0);
+
+    if (_nDims == 3) {
+        int max3DTexDim;
```
Comment (clyne) [inline lib/render/Texture.cpp:58]:
```suggestion
        GLint max3DTexDim;
```

File: lib/render/Texture.cpp [hunk 2]
```diff
@@ -52,6 +53,23 @@ int Texture::TexImage(int internalFormat, int width, int height, int depth, unsi
...
+        glGetIntegerv(GL_MAX_3D_TEXTURE_SIZE, &max3DTexDim);
+        if (width > max3DTexDim || height > max3DTexDim || depth > max3DTexDim) {
+            Wasp::MyBase::SetErrMsg("Texture size (%lix%lix%li) not supported by GPU (max supported size per dim is %i)\n", width, height, depth, max3DTexDim);
+            return -1;
+        }
+    } else {
+        int max2DTexDim;
```
Comment (clyne) [inline lib/render/Texture.cpp:65]:
```suggestion
        GLint max2DTexDim;
```

File: lib/render/glutil.cpp [hunk 3]
```diff
@@ -69,13 +69,57 @@ bool oglStatusOK(vector<int> &status)
...
+        msg += " (#" + std::to_string(e) + ")\n";
+    }
     return msg;
 }
 
+int oglGetFreeMemory()
+{
+#define TEXTURE_FREE_MEMORY_ATI 0x87FC
+#define GPU_MEMORY_INFO_CURRENT_AVAILABLE_VIDMEM_NVX 0x9049
+    int buf[4];
```
Comment (clyne) [inline lib/render/glutil.cpp:103]:
```suggestion
    GLint buf[4];
```

**Review by clyne — COMMENTED** [review]
Looks good. Just a couple minor nits to avoid compiler warnings.

**Review by clyne — APPROVED** [review]

File: lib/render/VolumeCellTraversal.cpp [hunk 4]
```diff
@@ -412,6 +412,26 @@ void VolumeCellTraversal::SetUniforms(const ShaderProgram *s) const
 
 float VolumeCellTraversal::GuestimateFastModeSpeedupFactor() const { return 2; }
 
+int VolumeCellTraversal::CheckHardwareSupport(const Grid *grid) const
+{
+    int ret = VolumeRegular::CheckHardwareSupport(grid);
+    if (ret < 0)
+        return ret;
+
+    long freeKB = oglGetFreeMemory();
```
Comment (shaomeng) [inline lib/render/VolumeCellTraversal.cpp:421]: The function oglGetFreeMemory() returns an int. So declare freeKB as int?

**Review by shaomeng — APPROVED** [review]

File: lib/render/VolumeCellTraversal.cpp (hunk 4 above)
Comment (sgpearse) [reply lib/render/VolumeCellTraversal.cpp:421]: freeKB will be promoted to a long later on in the code, so it's ok to just do it here.

**Pull request merged**